### ISSUE TITLE
`soft_margin_loss` gives wrong gradient when `target` with dtype uint8

### ISSUE THREAD

**Issue by upupyy** [pytorch/pytorch#80206]
### 🐛 Describe the bug

`soft_margin_loss`gives wrong gradient when`target` with dtype uint8

```python
import torch

torch.random.manual_seed(9782)
def get_fn():
    target = torch.tensor([0, 0, 1, 1], dtype=torch.uint8, device='cpu')
    def fn(input):
        fn_res = torch.nn.functional.soft_margin_loss(input, target)
        return fn_res
    return fn
fn = get_fn()
input = torch.tensor([[0.9707, 0.2672, 0.2389, 0.6876]], dtype=torch.float64, device='cpu', requires_grad=True)
try:
    torch.autograd.gradcheck(fn, (input), check_sparse_nnz=False, atol=0.01, rtol=0.01, check_forward_ad=False, check_backward_ad=True, check_batched_grad=False)
except Exception as e:
    print(e)

# Jacobian mismatch for output 0 with respect to input 0,
# numerical:tensor([[ 0.0000],
#         [ 0.0000],
#         [-0.1101],
#         [-0.0836]], dtype=torch.float64)
# analytical:tensor([[-0.0000],
#         [-0.0000],
#         [-0.2500],
#         [-0.2500]], dtype=torch.float64)
```

If `target` has dtype `int16` or `int32`, it will return the correct gradient.

### Versions

pytorch: 1.11.0
